DotNetHowToSecurityTrimming/DotNetHowToSecurityTrimming/Program.cs

```diff
@@ -0,0 +1,211 @@
+            Console.WriteLine("Refresh cache...\n");
+            RefreshCache();
+
+            // Create an HTTP reference to the catalog index
+            _searchClient = new SearchServiceClient(searchServiceName, new SearchCredentials(apiKey));
+            _indexClient = new SearchIndexClient(searchServiceName, indexName, new SearchCredentials(apiKey));
+
+            if (DeleteIndex(indexName))
+            {
+                Console.WriteLine("Creating index...\n");
+                CreateIndex(indexName);
+            }
+
+            Console.WriteLine("Indexing documents...\n");
+            // Index documents with relevant group ids
+            IndexDocuments(indexName, groups);
+
+            foreach (var user in UsersPrincipalNameGroup)
+            {
+                // Retrieve user's groups so that a search filter could be built using the groups list
+                Console.WriteLine("Get groups for user {0}...\n", user);
+                RefreshCacheIfRequired(user);
+                SearchQueryWithFilter(user);
+            }
+        }
+
+        private static async void RefreshCacheIfRequired(string user)
+        {
+            if (!_groupsCache.ContainsKey(user))
+            {
+                var groups = await _microsoftGraphHelper.GetGroupIdsForUser(user);
+                _groupsCache[user] = groups;
+            }
+        }
+        
+        private static async void RefreshCache()
+        {
+            HttpClient client = new HttpClient();
+            string responseString = await _microsoftGraphHelper.SendRequestAndGetResponse(client, UsersPrincipalNameGroup);
+
+            BatchResult data = JsonConvert.DeserializeObject<BatchResult>(responseString);
+            // Clear existing cache as new groups were retrieved
+            _groupsCache.Clear();
+            for (int i = 0; i < data.Responses.Count(); i++)
+            {
+                List<string> userGroups = new List<string>();
+                for (int j = 0; j < data.Responses[i].Body.Value.Count(); j++)
+                {
+                    string value = data.Responses[i].Body.Value[j];
+                    userGroups.Add(value);
+                }
+                int id = Convert.ToInt32(data.Responses[i].Id);
+                string key = UsersPrincipalNameGroup[id];
+                _groupsCache[key] = userGroups;
+            }
+        }
+
+        private static void SearchQueryWithFilter(string user)
+        {
+            SearchParameters parameters;
+            DocumentSearchResult<SecuredFiles> results;
+            // Using the filter below, the search result will contain all documents, that their GroupIds field contain any one of the 
+            // Ids in the groups list
+            string filter = String.Format("groupIds/any(p:search.in(p, '{0}'))", string.Join(",", String.Join(",", _groupsCache[user])));
+            parameters =
```

Comment:
you can inline parameter declaration here: SearchParameters parameters =

Reply:
@Yahnoosh I've addressed your comments, want to take another look?